components/renku_data_services/user_preferences/api.spec.yaml

```diff
@@ -0,0 +1,142 @@
+---
+openapi: 3.0.2
+info:
+  title: Renku user preferences services
+  description: |
+    Services that provide information about user preferences.
+  version: v1
+servers:
+  - url: /api/data
+  - url: /ui-server/api/data
+paths:
+  /user_preferences:
+    get:
+      summary: Get user preferences for the currently logged in user
+      responses:
+        "200":
+          description: The user preferences
+          content:
+            "application/json":
+              schema:
+                $ref: "#/components/schemas/UserPreferences"
+        "404":
+          description: The user has no preferences saved
+          content:
+            "application/json":
+              schema:
+                $ref: "#/components/schemas/ErrorResponse"
+        default:
+          $ref: "#/components/responses/Error"
+      tags:
+        - user_preferences
+  "/user_preferences/pinned_projects":
+    post:
+      summary: Add a pinned project
+      requestBody:
+        required: true
+        content:
+          "application/json":
+            schema:
+              $ref: "#/components/schemas/AddPinnedProject"
+      responses:
+        "200":
+          description: The updated user preferences
+          content:
+            "application/json":
+              schema:
+                $ref: "#/components/schemas/UserPreferences"
+        default:
+          $ref: "#/components/responses/Error"
+      tags:
+        - user_preferences
+    delete:
+      summary: Remove a pinned project
+      parameters:
+        - in: query
+          name: project_slug
+          required: true
+          schema:
+            type: string
+      responses:
+        "200":
+          description: The updated user preferences
+          content:
+            "application/json":
+              schema:
+                $ref: "#/components/schemas/UserPreferences"
+        default:
+          $ref: "#/components/responses/Error"
+      tags:
+        - user_preferences
+
+components:
+  schemas:
+    UserPreferences:
+      type: object
+      description: The object containing user preferences
+      additionalProperties: false
+      properties:
+        user_id:
+          $ref: "#/components/schemas/UserId"
+        pinned_projects:
+          $ref: "#/components/schemas/PinnedProjects"
+      required: ["user_id", "pinned_projects"]
+    UserId:
+      type: string
+      description: The unique identifier for a user
+      minLength: 3
+      example: "user-id-example"
+    PinnedProjects:
+      type: object
+      description: The list of projects a user has pinned on their dashboard
+      properties:
+        project_slugs:
+          type: array
+          items:
+            $ref: "#/components/schemas/ProjectSlug"
+    ProjectSlug:
+      type: string
+      description: The slug used to identify a project
+      minLength: 3
+      example: "user/my-project"
```

Comment:
This would be over-validation. I do not care if the project slug is valid or not, GitLab will give me the project data or not (in which case, the pin is ignored).

Reply:
Ok you have to handle errors or validation somewhere. I guess that will happen on the UI side.